docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+As described in [Issue#591](https://github.com/gardener/etcd-druid/issues/591) scheduling compaction only based on number of accumulated etcd events is not sufficient to ensure a successful compaction. This is specifically targeted for kubernetes clusters where each etcd event is larger in size owing to large spec or status fields or respective resources. 
+
+Druid will now need information regarding snapshot sizes, and more importantly the total size of accumulated delta snapshots since the last full snapshot.
+
+The authors propose to enhance the proposed `snapshots` field described in [Use Case #3](#use-case-3-monitoring-backup-health) with the following additional field:
+
+```yaml
+snapshots:
+  accumulatedDeltaSize: <total size of delta snapshots since last full snapshot>
+```
+
+Druid can then use this information in addition to the existing revision information to decide to trigger an early snapshot compaction job. This effectively allows druid to be proactive in performing regular compactions for etcds receiving large events, reducing the probability of a failed snapshot compaction or restoration.
+
+### Enhanced Defragmentation
+
+Reader is recommended to read [Etcd Compaction & Defragmentation](https://etcd.io/docs/v3.4/op-guide/maintenance/) in order to understand the following terminology:
+
+[`dbSize`](https://github.com/etcd-io/etcd/blob/a603c0798948941d453b158af794edab1a8230be/mvcc/backend/backend.go#L60-L64) - total storage space used by the etcd database
+
+[`dbSizeInUse`](https://github.com/etcd-io/etcd/blob/a603c0798948941d453b158af794edab1a8230be/mvcc/backend/backend.go#L65-L68) - logical storage space used by the etcd database, not accounting for free pages in the DB due to etcd history compaction
+
+The leading-backup-sidecar performs periodic defragmentations of the DBs of all the etcd-members in the cluster, controlled via a defragmentation cron schedule provided to each backup-sidecar. Defragmentation is a costly maintenance operation and causes a brief downtime to the etcd-member being defragmented, due to which the leading-backup-sidecar defragments each etcd-member sequentially. This ensures that only one etcd-member would be unavailable at any given time, thus avoiding an accidental quorum loss in the etcd cluster.
+
+The authors propose to move the responsibility of orchestrating these individual defragmentations to druid due to the following reasons:
+
+- Since each backup-sidecar only has knowledge of the health of its own etcd, it can only determine whether its own etcd can be defragmented or not, based on etcd-member health. Trying to defragment a different healthy etcd-member while another etcd-member is unhealthy would lead to a transient quorum loss.
+- Each backup-sidecar is only a `sidecar` to its own etcd-member, and by good design principles, it must not be performing any cluster-wide maintenance operations, and this responsibility should remain with the etcd cluster operator.
```

Comment:
>  and by good design principles, it must not be performing any cluster-wide maintenance operations, and this responsibility should remain with the etcd cluster operator.

AFAIU, you are suggesting that defragmentation should be triggered by etcd-druid itself ... Is it ?
But previous statement mentioned etcd-druid should be just orchestrate the defragmention. Please correct me if I understood in wrong way.

Reply:
Defragmentation will be triggered, ie initiated by druid. This means that druid issues a defragmentation call to a particular etcd member. The member will then perform the actual defragmentation by calling the defragment API on the etcd.

So while druid is still not directly talking to any etcd process, it still decides the order in which to defragment the members, and "orchestrates" the defragmentation.

Reply:
k8s native way would be for druid to signal intent to etcd-member via well defined API objects, which then reacts to the requested operation and updates the status of the said API object (in this case that would be `EtcdMember`). So there is no direct communication between druid and any etcd-member.

Reply:
Yes. That's one of the main objectives of this proposal, to provide a way where each etcd member can reliably publish its state to etcd-druid. Etcd-druid will be able to derive each member's health via the pod's readiness probe check, the liveness heartbeats via each member lease, and also the individual EtcdMember state transitions (to determine the state and subState of each member). By seeing the `defragmentations`, `dbSize` and `dbSizeInUse` fields maintained in the EtcdMember, druid will additionally know the status of previous defragmentations as well as triggers on when the next defragmentation should be performed.